homeassistant/components/legrandinone.py

```diff
@@ -0,0 +1,505 @@
+
+        # Process the initial event now that the entity is created
+        if self._initial_event:
+            self.handle_event_callback(self._initial_event)
+
+
+class LegrandInOneCommand(LegrandInOneDevice):
+    """Singleton class to make IOBL command interface available to entities.
+
+    This class is to be inherited by every Entity class that is actionable
+    (switches/lights). It exposes the IOBL command interface for these
+    entities.
+
+    The IOBL interface is managed as a class level and set during setup (and
+    reset on reconnect).
+    """
+
+    _protocol = None
+
+    @classmethod
+    def set_iobl_protocol(cls, protocol):
+        """Set the IOBL asyncio protocol as a class variable."""
+        cls._protocol = protocol
+
+    @classmethod
+    def is_connected(cls):
+        """Return connection status."""
+        return bool(cls._protocol)
+
+    @classmethod
+    async def send_command(cls, command_data):
+        """Send device command to IOBL."""
+        return await cls._protocol.send_packet(command_data)
+
+    async def _async_handle_command(self, command, *args):
+
+        if command == 'turn_on':
+            cmd = 'on'
+            self._state = True
+
+        elif command == 'turn_off':
+            cmd = 'off'
+            self._state = False
+
+        elif command == 'dim':
+            # convert brightness to rflink dim level
+            cmd = int(args[0] * 100 / 255)
+            self._state = True
+
+        elif command == 'toggle':
+            cmd = 'on'
+            # if the state is unknown or false, it gets set as true
+            # if the state is true, it gets set as false
+            self._state = self._state in [None, False]
+
+        # Cover options for IOBL
+        elif command == 'close_cover':
+            cmd = 'move_down'
+            self._state = False
+
+        elif command == 'open_cover':
+            cmd = 'move_up'
+            self._state = True
+
+        elif command == 'stop_cover':
+            cmd = 'move_stop'
+            self._state = True
+
+        # Send initial command and queue repetitions.
+        # This allows the entity state to be updated quickly and not having to
+        # wait for all repetitions to be sent
+        await self._async_send_command(cmd)
+
+        # Update state of entity
+        await self.async_update_ha_state()
+
+    async def _async_send_command(self, cmd):
+        """Send a command for device to iobl gateway."""
+        _LOGGER.debug(
+            "Sending command: %s to iobl device: %s", cmd, self.legrand_id)
+
+        if not self.is_connected():
+            raise HomeAssistantError('Cannot send command, not connected!')
+
+        # Puts command on outgoing buffer and returns straight away.
+        # IOBL protocol/transport handles asynchronous writing of buffer
+        # to serial/tcp device. Does not wait for command send
+        # confirmation.
+        data = cast(Dict[str, Any], {
```

Comment:
Home assistant shouldn't need to be aware of the protocol packet structure. This should be abstracted by the interface library.

Reply:
I'm not quite sure to understand the remark here. What should be abstracted ?

Reply:
Well, I'm not sure here. The thing is, the library allows communication on a "bus", and then each component in home assistant will know how to command, using this "bus" and its provided interfaces, the specific type of devices corresponding to that specific platform. I carefully looked at the hlk_sw16 platform, but here things are a little different. Things are more like rflink or rfxtrx or KNX.